xmltar.1
.TH XMLTAR "1" "June 2010" "xmltar 0.2" "User Commands"
.SH NAME
xmltar \- manual page for xmltar 0.2
.SH SYNOPSIS
.B xmltar
[\fIOPTION\fR...] [\fIFILE\fR]...
.SH DESCRIPTION
`xmltar' saves many files together into a disk archive, and can
restore individual files from the archive.
.SH EXAMPLES
.TP
xmltar \fB\-cf\fR archive.xmltar foo bar
# Create archive.xmltar from files foo and bar.
.TP
xmltar \fB\-xf\fR archive.xmltar
.TP
xmltar \fBcf\fR 
# Extract all files from archive.xmltar.
.TP
xmltar \fBcf\fR archive-%03d.xmltar
\fB--multi-volume\fR \
\fB--starting-volume=1\fR \
\fB--tape-length=940000000\fR \
\fB--stop-after 5\fR \
\fB--listed-incremental archive.xmltar.list\fR \
\fB--xz\fR \
\fRarchive.xmltar
.br
# Create a series of multivolume archives with names following
the pattern archive.001.xmltar, archive.002.xmltar, ..., where
each archive is exactly 940Mb long. Stop adding additional files
to the archive after 5 volumes have been completed. Use xz to
compress the archive. Create a list of files which have been
archive in the file archive.xmltar.list.
.IP
.SH OPTIONS
.SH Main operation mode:
.TP
\fB\-c\fR, \fB\-\-create\fR
create a new archive
.TP
\fB\-r\fR, \fB\-\-append\fR
append files to the end of an archive
.TP
\fB\-t\fR, \fB\-\-list\fR
list the contents of an archive
.TP
\fB\-x\fR, \fB\-\-extract\fR, \fB\-\-get\fR
extract files from an archive
.IP
.SH\fBCommon Options:\fR
.TP
\fB-f\fR, \fB--file=ARCHIVE\fR
use archive file or device ARCHIVE
.TP
\fB-j\fR, \fB--bzip2\fR
filter the archive through bzip2
.TP
\fB -J\fR, \fB--xz\fR
filter the archive through xz
.TP
\fB-z\fR, \fB--gzip\fR
filter the archive through gzip

.SH Operation modifiers:
.TP
\fB\-M\fR, \fB\-\-multi\-volume\fR
create/list/extract multi\-volume archive
.TP
\fB\-j\fR, \fB\-\-bzip2\fR
filter the archive through bzip2
.TP
\fB\-\-lzma\fR
filter the archive through lzma
.TP
\fB\-z\fR, \fB\-\-gzip\fR, \fB\-\-gunzip\fR, \fB\-\-ungzip\fR
filter the archive through gzip
.TP
\fB\-J\fR, \fB\-\-xz\fR
filter the archive through xz
.TP
\fB\-v\fR, \fB\-\-verbose\fR
verbosely list files processed
.SH Other options:
.TP
\-?, \fB\-\-help\fR
give this help list
.SH AUTHOR
Written by David A. Betz.
.SH "REPORTING BUGS"
Report bugs to <dbetz@sbcglobal.net>.
.SH COPYRIGHT
Copyright \(co 2010 David A. Betz
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
Command line options for xmltar are meant to be similar
to the command line options for
.B tar\fR.


CHANGELOG
commit e4b014b5386a070fbacbaf12c19cb32ef40f38e1
Author: David Betz <dbetz@www.memorial-anesthesia.org>
Date:   Sun Jun 6 06:06:19 2010 -0700

    Make better use of blocking IO to hopefully increase throughput.

commit 008605b513a31dde30de9280b5b3d7412b49dbd1
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Fri Jun 4 17:01:03 2010 -0700

    Moved member function bodies into .cpp

commit fb4545ce0cc1d482773084cf92feeece550102dc
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Wed Jun 2 05:39:47 2010 -0700

    initialized all pipe fds to -1 so that we don't try to close unopened pipefds.
    checked pipe_state before closing, so that if pipe_state_ is unopened, we don't try to close any fds.

commit b304a9ddbd74a383ebb7caa6151f04535dae7a73
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Mon May 31 07:41:43 2010 -0700

    Release 0.1

commit a8b5994ab8dfb139a7ef0039fed0ed7c8aeeed87
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Mon May 31 05:42:42 2010 -0700

    Added GPL license information to all source files.

commit ca7b631c402f3f38dfcdbcbb1a3a96fc185755f2
Author: David Betz <dbetz@localhost.localdomain>
Date:   Sun May 30 16:40:52 2010 -0700

    Added XmltarOptions, which contains all xmltar options and also
    common subroutines for finding proper encoders/decoders.

commit aac8909a57a69335f63f0025faebf7c19def0cff
Author: David Betz <dbetz@localhost.localdomain>
Date:   Sun May 30 16:39:32 2010 -0700

    Added guards to tabs and newlines.
    Added --no-tabs and --no-newlines options.
    Factored out options from Xmltar, Archive, and Archive_Member.

commit 413603b95b9d3e74691545ee968d49d6391b9b4e
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sun May 30 06:31:01 2010 -0700

    implemented base64 encoding

commit d58bafb55defe3fdb901d7d61e592371668f8e05
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Wed May 26 05:36:58 2010 -0700

    Rearrangement of extract parser.

commit 98ba7682da5d7c40db57b8ba5fa2758e56724ba4
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Tue May 25 15:21:17 2010 -0700

    Corrected bug in Bidirectional_Pipe::Initialize. It is necessary to set the FD_CLOEXEC flags on
    the pipe file descriptors. Otherwise, each fork after the first will duplicate *all* preceding file
    descriptors, leaving multiple writers to the slave pipes, resulting in the slave pipes never
    seeing EOF because they have writers. Investigated using the tool 'lsof -a -p <pid>' and
    'find -type p'.

commit ae7fae7ed24249df8ee2a186ff09d25bff168274
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sun May 23 09:22:42 2010 -0700

    - unguarded debugging statements removed
    - before eliminating special case identity and base16 encoders

commit 2aaa446aa54398b1ff9f037850db416c1dac3b9a
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sat May 22 18:22:34 2010 -0700

    corrected bug in basic_pipe_transitbuf::underflow

commit 4c10eaa0ff45743d363791ac363fdbc955670795
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Fri May 21 07:13:49 2010 -0700

    Extract can now extract multivolume archives

commit 916e910f029471fc881b9dbc58ae5ee51352caeb
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Tue May 18 06:06:33 2010 -0700

    Spirit extract works for directories and links.

commit 526a6b4f4e7ff0a435f3a5ad40ba67350330fb62
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sun May 2 09:14:39 2010 -0700

    xz, bzip2, gz, identity all implemented

commit 20c75c7cb36e6e7f8ac7546713fb99bd1b5d533e
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Thu Apr 29 20:22:44 2010 -0700

    Identity postcompress multivolume working.

commit 7ee7448a0cb2cdf8ba8a84bed9e0cd955cc5a93d
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Mon Apr 26 05:45:01 2010 -0700

    Rearrangement of source files.
    Consistent naming of header and trailer for Archive and Archive_Member

commit 2fcad4b0bede8e62fb2c444bf1d84b01fb6cbe2f
Author: dbetz <dbetz@localhost>
Date:   Thu Apr 22 14:28:54 2010 -0700

    hunting down bug in Open_New_Postcompressor

commit db5d66ba3c632a7a45235e1a14fd58349ca0e865
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Thu Apr 22 06:11:48 2010 -0700

    basic_pipe_transitbuf_base created, but not working correctly.

commit fc3857bb14c8b722c2c7b3c9529b2c76727b8484
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Wed Apr 21 06:02:39 2010 -0700

    Collapsing of Bidirectional_Pipe::Close using default arguments.
    Added back identitybuf.hpp

commit e04f6bbe7599f5a4d4a32fff4e2ca6ef7fab8708
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Tue Apr 20 06:08:57 2010 -0700

    Codec implemented, almost working on simple archives

commit 7148ba46f2192034edc3226eaf21e7d8133f9bcc
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Mon Apr 19 06:15:31 2010 -0700

    Bidirectional_Pipe::close_write now empties the write buffer before closing the pipe.
    Began work on Codec class as a base class of all encoders/decoders.

commit e0c77a7a15170d188a4c68ea4172cf79f5fbbdb7
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sun Apr 18 11:38:59 2010 -0700

    Works for single-volume archives.

commit 6631c64cd13ccb6c2d1e8e8d5b04648e3794029b
Author: David Betz <dbetz@www.memorial-anesthesia.org>
Date:   Sat Apr 17 19:33:53 2010 -0700

    added back .cproject and .project

commit bde2ca9a43edff3234f1b742fd042214b2c6b4fe
Author: David Betz <dbetz@www.memorial-anesthesia.org>
Date:   Sat Apr 17 19:27:54 2010 -0700

    deleted .cproject and .project

commit 9f04ae81c4ff2548ed6e92ebaa8e9dd3ca5979b8
Author: David Betz <dbetz@www.memorial-anesthesia.org>
Date:   Sat Apr 17 17:20:37 2010 -0700

    Rearranged files for Eclipse

commit 674b93a1b5066c9804c799a1b75ad610d65b4be3
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sat Apr 17 17:17:46 2010 -0700

    Xmltar5

commit 266f58b5184e25c7ee3002bd40beb3001293c46b
Author: dbetz <dbetz@memorial-anesthesia.org>
Date:   Sat Apr 17 07:01:06 2010 -0700


src/common/Xmltar/Xmltar.cpp
/*

Copyright 2010 by David A. Betz

This file is part of xmltar.

xmltar is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

xmltar is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with xmltar.  If not, see <http://www.gnu.org/licenses/>.

*/

#include "Xmltar/Xmltar.hpp"

#include <limits>
#include "Tar_Options.hpp"

Xmltar::Xmltar(int argc, char *argv[])
	: version("Xmltar_0_0_1") {
	Parse_Opts::Option_Parser p;

	p.Add_Option(Parse_Opts::ARGS_0,"-c","--create","create archive",
			p.Assign_Value(this->options_.operation,XmltarOptions::CREATE));
	p.Add_Option(Parse_Opts::ARGS_0,"-r","--append","append to existing archive",
			p.Assign_Value(this->options_.operation,XmltarOptions::APPEND));
	p.Add_Option(Parse_Opts::ARGS_0,"-t","--list","list table of contents of existing archive",
			p.Assign_Value(this->options_.operation,XmltarOptions::LIST));
	p.Add_Option(Parse_Opts::ARGS_0,"-x","--extract","extract files from an archive",
			p.Assign_Value(this->options_.operation,XmltarOptions::EXTRACT));
	p.Add_Option(Parse_Opts::ARGS_0,"-v","--verbose","report more details",
			p.Increment_Value(this->options_.verbosity));

    p.Add_Option(Parse_Opts::ARGS_0,"-z","--gzip","compress archive",
            p.Assign_Value(this->options_.postcompress, XmltarOptions::GZIP));
    p.Add_Option(Parse_Opts::ARGS_0,"","--bzip2","compress archive",
            p.Assign_Value(this->options_.postcompress, XmltarOptions::BZIP2));
    p.Add_Option(Parse_Opts::ARGS_0,"","--xz","compress archive",
            p.Assign_Value(this->options_.postcompress, XmltarOptions::XZ));

    p.Add_Option(Parse_Opts::ARGS_0,"","--pre-gzip","pre-compress files before archiving",
            p.Assign_Value(this->options_.precompress, XmltarOptions::GZIP));
    p.Add_Option(Parse_Opts::ARGS_0,"","--pre-bzip2","pre-compress files before archiving",
            p.Assign_Value(this->options_.precompress, XmltarOptions::BZIP2));
    p.Add_Option(Parse_Opts::ARGS_0,"","--pre-xz","pre-compress files before archiving",
            p.Assign_Value(this->options_.precompress, XmltarOptions::XZ));

    p.Add_Option(Parse_Opts::ARGS_0,"","--base16","base16 encode files before archiving",
            p.Assign_Value(this->options_.encoding,XmltarOptions::BASE16));
    p.Add_Option(Parse_Opts::ARGS_0,"","--base64","base64 encode files before archiving",
            p.Assign_Value(this->options_.encoding,XmltarOptions::BASE64));

	p.Add_Option(Parse_Opts::ARGS_0,  "","--compress-listed-incremental","compress listed-incremental file",
			p.Assign_Value(this->options_.compress_listed_incremental_file,true));
	p.Add_Option(Parse_Opts::ARGS_1,"-f","--file","specify archive to create/append/list/extract",
			p.Assign_Args(this->options_.base_xmltar_file_name));
	p.Add_Option(Parse_Opts::ARGS_1,"","--exclude","specify PATTERN to exclude files from archiving",
			p.Append_Args(this->options_.exclude_files));
	p.Add_Option(Parse_Opts::ARGS_1,"-g","--listed-incremental","work with listed-incremental archives",
			p.Assign_Args(this->options_.listed_incremental_file),
			p.Assign_Value(this->options_.incremental,true));
	p.Add_Option(Parse_Opts::ARGS_0,"-M","--multi-volume","work with multivolume archives",
			p.Assign_Value(this->options_.multi_volume,true));
	bool tape_length_option=false;
	p.Add_Option(Parse_Opts::ARGS_1,"-L","--tape-length","specify size of a multivolume archive",
            p.Assign_Args(this->options_.tape_length),
            p.Assign_Value(tape_length_option,true));
	p.Add_Option(Parse_Opts::ARGS_1,"","--starting-volume","starting volume number of a multivolume archive",
			p.Assign_Args(this->options_.starting_sequence_number));
	p.Add_Option(Parse_Opts::ARGS_1,"-T","--files-from","specify source of files to archive",
			p.Assign_Args(this->options_.files_from));
    p.Add_Option(Parse_Opts::ARGS_1,"","--stop-after","stop archiving additional files once n volumes have been created",
            p.Assign_Args(this->options_.stop_after));
    p.Add_Option(Parse_Opts::ARGS_1,"","--tag","tag archived files with identifier",
            p.Assign_Args(this->options_.tag));
	p.Add_Option(Parse_Opts::ARGS_0,"","--no-tabs","do not tab-indent the xml archive",
			p.Assign_Value(this->options_.tabs_,false));
	p.Add_Option(Parse_Opts::ARGS_0,"","--no-newlines","do not include any newlines in the xml archive",
			p.Assign_Value(this->options_.newlines_,false));

	std::vector<std::string> remaining_args=p.Parse(argc,argv);
	for(std::vector<std::string>::iterator i=remaining_args.begin(); i!=remaining_args.end(); ++i)
		options_.source_files.push_back(boost::filesystem::path(*i));

	options_.current_sequence_number=options_.starting_sequence_number;

	if (options_.verbosity==3){
		std::cerr << "Argc=" << argc << std::endl;
		if (options_.operation==XmltarOptions::APPEND) std::cerr << "Operation=APPEND" << std::endl;
		if (options_.operation==XmltarOptions::CREATE) std::cerr << "Operation=CREATE" << std::endl;
		if (options_.operation==XmltarOptions::LIST) std::cerr << "Operation=LIST" << std::endl;
		if (options_.operation==XmltarOptions::EXTRACT) std::cerr << "Operation=EXTRACT" << std::endl;
		std::cerr << "verbosity=" << options_.verbosity << std::endl;
		if (options_.multi_volume) std::cerr << "Multivolume" << std::endl;
		std::cerr << "starting_sequence_number=" << options_.starting_sequence_number << std::endl;
		std::cerr << "listed-incremental=" << options_.incremental << std::endl;
		std::cerr << "tape length=" << options_.tape_length << std::endl;
		std::cerr << "stop after=" << options_.stop_after << std::endl;
		std::cerr << "base_xmltar_file_name=" << options_.base_xmltar_file_name << std::endl;
		std::cerr << "Source file size=" << options_.source_files.size() << std::endl;
		for(std::vector<boost::filesystem::path>::iterator i=options_.exclude_files.begin(); i!=options_.exclude_files.end(); ++i)
			std::cerr << "Exclude file=" << *i << std::endl;
		for(std::vector<boost::filesystem::path>::iterator i=options_.source_files.begin(); i!=options_.source_files.end(); ++i)
			std::cerr << "Source file=" << *i << std::endl;
		std::cerr << "listed-incremental file=" << options_.listed_incremental_file << std::endl;
		std::cerr << "files from=" << options_.files_from << std::endl;
	}

	if (tape_length_option && !options_.multi_volume)
	    throw "Xmltar::Xmltar: tape_length specification requires multivolume specification (to indicate overflow archive name)";
}
